File: engine/src/Projectile.ts
// 投射物：远程/法术/攻城单位的飞行物。瞬间发射 → 飞行 → 命中目标
// Unity 迁移：MonoBehaviour + Transform.position 插值；或 ECS 组件
// 当前实现：直线插值，命中时锁定目标位置（追踪式）；目标死亡 → 撞已死位置 → 自销毁

import { Vec2 } from "./Vec2.ts";
import type { Side, DamageType } from "./Enums.ts";
import type { Unit } from "./Unit.ts";
import type { Building } from "./Building.ts";

export type ProjectileTarget = Unit | Building;

export class Projectile {
  static _nextId = 1;

  readonly id: number;
  readonly side: Side;
  pos: Vec2;
  readonly target: ProjectileTarget;
  readonly damage: number;
  readonly damageType: DamageType;
  readonly speed: number;          // px/sec (世界坐标)
  readonly aoeRadius: number;      // 0 = 单体
  readonly visualKind: "arrow" | "bolt" | "shell" | "magic";

  alive = true;
  hitAt: number | null = null;     // 命中时间戳（用于动画延迟销毁）

  constructor(opts: {
    side: Side;
    pos: Vec2;
    target: ProjectileTarget;
    damage: number;
    damageType: DamageType;
    speed: number;
    aoeRadius?: number;
    visualKind?: "arrow" | "bolt" | "shell" | "magic";
  }) {
    this.id = Projectile._nextId++;
    this.side = opts.side;
    this.pos = opts.pos.clone();
    this.target = opts.target;
    this.damage = opts.damage;
    this.damageType = opts.damageType;
    this.speed = opts.speed;
    this.aoeRadius = opts.aoeRadius ?? 0;
    this.visualKind = opts.visualKind ?? "arrow";
  }

  /** 推进一帧；返回 true 表示命中（外部应用伤害+销毁） */
  step(dt: number): boolean {
    if (!this.alive) return false;
    const tx = this.target.pos.x;
    const ty = this.target.pos.y;
    const dx = tx - this.pos.x;
    const dy = ty - this.pos.y;
    const dist = Math.sqrt(dx * dx + dy * dy);
    const stepLen = this.speed * dt;
    if (dist <= stepLen + 4) {
      this.pos = new Vec2(tx, ty);
      return true;
    }
    this.pos = new Vec2(this.pos.x + (dx / dist) * stepLen, this.pos.y + (dy / dist) * stepLen);
    return false;
  }
}

/** 根据单位 role / damageType 推断默认投射物视觉 + 飞行速度 */
export function inferProjectile(role: string, damageType: DamageType): { visualKind: "arrow" | "bolt" | "shell" | "magic"; speed: number } {
  if (role === "siege") return { visualKind: "shell", speed: 700 };
  if (role === "caster" || damageType === "magic") return { visualKind: "magic", speed: 900 };
  if (role === "range" || damageType === "pierce") return { visualKind: "arrow", speed: 1100 };
  return { visualKind: "bolt", speed: 1000 };
}

/** 远程判定：range > 200 视为投射物单位（近战不走飞行）*/
export function isRangedAttack(rangePx: number): boolean {
  return rangePx > 200;
}

File: unity/Assets/Mg01/Sim/Projectile.cs
// 与 engine/src/Projectile.ts 一一对应

using UnityEngine;
using Mg01.Core;

namespace Mg01.Sim
{
    public enum ProjectileVisual { Arrow, Bolt, Shell, Magic }

    public class Projectile : MonoBehaviour
    {
        public static int _nextId = 1;

        public int id;
        public Side side;
        public Object target;       // Unit 或 Building
        public float damage;
        public DamageType damageType;
        public float speed;          // px/sec (世界坐标)
        public float aoeRadius;      // 0 = 单体
        public ProjectileVisual visualKind = ProjectileVisual.Arrow;

        public bool alive = true;
        public float? hitAt;         // 命中时间戳

        public Vector2 Pos => transform.position;

        public void Init(Side s, Vector2 startPos, Object tgt, float dmg, DamageType dt,
                         float spd, float aoe = 0f, ProjectileVisual vk = ProjectileVisual.Arrow)
        {
            id = _nextId++;
            side = s;
            transform.position = startPos;
            target = tgt;
            damage = dmg;
            damageType = dt;
            speed = spd;
            aoeRadius = aoe;
            visualKind = vk;
        }

        /// <summary>推进一帧；返回 true = 命中</summary>
        public bool Step(float dt)
        {
            if (!alive) return false;
            Vector2 tpos = TargetPos();
            Vector2 me = Pos;
            float dx = tpos.x - me.x;
            float dy = tpos.y - me.y;
            float dist = Mathf.Sqrt(dx * dx + dy * dy);
            float stepLen = speed * dt;
            if (dist <= stepLen + 4f)
            {
                transform.position = tpos;
                return true;
            }
            transform.position = new Vector2(me.x + (dx / dist) * stepLen, me.y + (dy / dist) * stepLen);
            return false;
        }

        private Vector2 TargetPos()
        {
            if (target is Unit u) return u.Pos;
            if (target is Building b) return b.Pos;
            return Pos;
        }

        // ---- 与 TS inferProjectile / isRangedAttack 对应 ----
        public static (ProjectileVisual visual, float speed) Infer(string role, DamageType damageType)
        {
            if (role == "siege")  return (ProjectileVisual.Shell, 700f);
            if (role == "caster" || damageType == DamageType.Magic) return (ProjectileVisual.Magic, 900f);
            if (role == "range"  || damageType == DamageType.Pierce) return (ProjectileVisual.Arrow, 1100f);
            return (ProjectileVisual.Bolt, 1000f);
        }

        public static bool IsRangedAttack(float rangePx) => rangePx > 200f;
    }
}
